### PR TITLE
Save waitTime value for sequence and conductor action (allow parent/child transaction ids)

### PR DESCRIPTION
**Pull request by upgle**
## Description
Currently, The openwhisk uses the `waitTime` annotation to store the internal wait time until the action code is provisioned. But this value is only present for non-sequence related activations.

Internally invoked actions can't know `waitTime` because this value is measured based on the start value of the first created TransactionId by topmost action.

So those actions invoked by topmost action must have a different TransactionId with a different start value.

To solve this problem I've modified the TransactionId case class to have a parent/child structure.

### Define the starting point for internally invoked action

This is the invocation flow of the sequence and conductor action in the controller. 
(Each action calls internal actions in a different way, so the starting point is different.)

#### Sequence action
An `invokeNextAction` method could be the starting point to create child transactionId.

image

```
wsk action create sequence --sequence increment,increment,increment
```
```
[05:04:08.704Z] [#tid_cWCsYNy] [ActionsApi] invokeAction sequence
[05:04:08.705Z] [#tid_cWCsYNy] [ActionsApi] invokeNextAction increment <- create child
[05:04:08.705Z] [#tid_cWCsYNy] [ActionsApi] invokeAction increment
[05:04:08.705Z] [#tid_cWCsYNy] [ActionsApi] invokeSingleAction increment
[05:04:08.705Z] [#tid_cWCsYNy] [ActionsApi] invokeSimpleAction increment
[05:04:08.807Z] [#tid_cWCsYNy] [ActionsApi] invokeNextAction increment  <- create child
[05:04:08.807Z] [#tid_cWCsYNy] [ActionsApi] invokeAction increment
[05:04:08.807Z] [#tid_cWCsYNy] [ActionsApi] invokeSingleAction increment
[05:04:08.807Z] [#tid_cWCsYNy] [ActionsApi] invokeSimpleAction increment
[05:04:08.828Z] [#tid_cWCsYNy] [ActionsApi] invokeNextAction increment  <- create child
[05:04:08.828Z] [#tid_cWCsYNy] [ActionsApi] invokeAction increment
[05:04:08.828Z] [#tid_cWCsYNy] [ActionsApi] invokeSingleAction increment
[05:04:08.828Z] [#tid_cWCsYNy] [ActionsApi] invokeSimpleAction increment
```

#### Conductor action
A `tryInvokeNext` and `invokeConductor` method could be the starting point to create child transactionId.

image

```
[05:05:55.575Z] [#tid_oVi9Qxc] [ActionsApi] invokeAction tripleAndIncrement
[05:05:55.575Z] [#tid_oVi9Qxc] [ActionsApi] invokeSingleAction tripleAndIncrement
[05:05:55.575Z] [#tid_oVi9Qxc] [ActionsApi] invokeComposition tripleAndIncrement
[05:05:55.577Z] [#tid_oVi9Qxc] [ActionsApi] invokeConductor Some({})  <- create child
[05:05:55.578Z] [#tid_oVi9Qxc] [ActionsApi] ┗ invokeSimpleAction tripleAndIncrement
[05:05:55.720Z] [#tid_oVi9Qxc] [ActionsApi] ┗ tryInvokeNext whisk.system/triple  <- create child
[05:05:55.728Z] [#tid_oVi9Qxc] [ActionsApi]    ┗ invokeComponent triple
[05:05:55.728Z] [#tid_oVi9Qxc] [ActionsApi]        ┗ invokeSimpleAction triple
[05:05:55.879Z] [#tid_oVi9Qxc] [ActionsApi] invokeConductor Some({"value":null})  <- create child
[05:05:55.880Z] [#tid_oVi9Qxc] [ActionsApi] ┗ invokeSimpleAction tripleAndIncrement
[05:05:55.897Z] [#tid_oVi9Qxc] [ActionsApi] ┗ tryInvokeNext whisk.system/increment  <- create child
[05:05:55.902Z] [#tid_oVi9Qxc] [ActionsApi]   ┗ invokeComponent increment
[05:05:55.902Z] [#tid_oVi9Qxc] [ActionsApi]       ┗ invokeSimpleAction increment
[05:05:56.771Z] [#tid_oVi9Qxc] [ActionsApi] invokeConductor Some({"value":1})  <- create child
[05:05:56.771Z] [#tid_oVi9Qxc] [ActionsApi] ┗ invokeSimpleAction tripleAndIncrement
```

### Creating a new child TransactionId
You can create a new child transaction by calling the childOf method of TransactionId.

```scala
  private def tryInvokeNext(user: Identity,
                            fqn: FullyQualifiedEntityName,
                            params: Option[JsObject],
                            session: Session,
                            parentTid: TransactionId): Future[ActivationResponse] = {

    implicit val transid: TransactionId = TransactionId.childOf(parentTid)
```

### Print root transaction id and leaf transaction id together
image

## Related issue and scope
https://github.com/apache/openwhisk/issues/3083

## My changes affect the following components
- TransactionId
- Intrinsic actions (e.g., sequences, conductors)
- Documentation

## Types of changes
- Enhancement or new feature (adds new functionality).
- Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [x] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

**Comment by rabbah** [comment]
Awesome! Thank you for adding this.

**Comment by upgle** [comment]
I've found a performance improvement point for this PR. As you know, using the string builder is much faster than the current version.

The blue line shows the performance of the transaction id generator currently used by openwhisk, and the orange line shows the performance of the implementation using the string builder. 

 I'll add a new commit.

image

(In the chart, the x-axis is the number of repetitions and the y-axis is the time taken.)

### Source code for performance testing

```scala

object TransactionId {
  private val dict = ('A' to 'Z') ++ ('a' to 'z') ++ ('0' to '9')

  def generateTidByMap(): String = {
    (0 until 32).map(_ => dict(ThreadLocalRandom.current().nextInt(dict.size))).mkString("")
  }

  def generateTidByStringBuilder(): String = {
    val sb = new StringBuilder
    for (_ <- 1 to 32) {
      sb.append(dict(util.Random.nextInt(dict.length)))
    }
    sb.toString
  }
}
```

File: common/scala/src/main/scala/org/apache/openwhisk/common/TransactionId.scala [hunk 1]
```diff
@@ -172,13 +175,22 @@ case class TransactionId private (meta: TransactionMetadata) extends AnyVal {
...
    * @param marker: The marker to add to the message
    */
   private def createMessageWithMarker(message: String, marker: LogMarker): String = s"$message $marker"
+
+  /**
+   * Find root transaction metadata
+   */
+  @tailrec
+  private def findRoot(meta: TransactionMetadata): TransactionMetadata =
+    if (meta.parent.isDefined) findRoot(meta.parent.get) else meta
```
Comment (rabbah) [inline common/scala/src/main/scala/org/apache/openwhisk/common/TransactionId.scala]: @markusthoemmes tailrec work with option map?

Comment (rabbah) [inline common/scala/src/main/scala/org/apache/openwhisk/common/TransactionId.scala]:
```suggestion
    meta.parent match {
       case Some(parent) => findRoot(parent)
       case _ => meta
```

Comment (rabbah) [reply common/scala/src/main/scala/org/apache/openwhisk/common/TransactionId.scala]: avoid `.get` even though it's safe, also clearer to me

**Comment by style95** [comment]
@upgle Sorry for letting this become stale.
Could you address the conflicts?

File: common/scala/src/main/scala/org/apache/openwhisk/common/TransactionId.scala (hunk 1 above)
Comment (upgle) [reply common/scala/src/main/scala/org/apache/openwhisk/common/TransactionId.scala]: @rabbah Thank you for your.sample code, I've modified the code.

**Comment by upgle** [comment]
@style95 @rabbah 
I've updated this PR, and all test cases have been passed. Please review it again.

**Pull request merged**